### PR TITLE
change transformMPI to pass Communicator instead of PromiseGuard

### PR DESCRIPTION
**Pull request by albestro**
This PR aims at proposing a way to make internal communication kernels deal with `Communicator`s instead of `PromiseGuard<Communicator>`s.

Let's consider the internal reduce communication kernel as an example:
https://github.com/eth-cscs/DLA-Future/blob/5a23516e72a2cff20a8672e5a5bb8f71809ae092/include/dlaf/communication/kernels/reduce.h#L40-L48

which currently accepts as parameter a `PromiseGuard<Communicator>`.

For what concerns the kernel implementation, it is interesting just the `Communicator`, but currently it has the burden of unwrapping it from the `PromiseGuard` that contains it before using it.

Apart from the (minor) burden of the unwrapping, passing a `PromiseGuard` implicitly enforces a superfluous constraint: in order to use it, we cannot use a `Communicator` as it is, but we have to wrap it inside a `Pipeline`.

We may want to drop this requirement, since we may soon have use-cases with algorithms that do not require ordered communications, i.e. by means of a `Pipeline`, which might also represent an over-constraint to the parallelisation.

After a quick chat with @msimberg about this, we thought about exploring the solution of adapting `transformMPI` so that it still "consumes" a `PromiseGuard` just after the kernel finishes (i.e. waiting just for the posting, not for its completion), and at the same time it also unwraps the `PromiseGuard<Communicator>` by calling `.ref()`. This is a "conditional" management, which means that if the `Communicator` is passed as it is without a `PromiseGuard`, it will go through and it won't get altered.

There might be drawbacks, and this PR is for sharing with the team so that we can find and implement the best solution.

Looking forward to your comments/feedbacks!

### CODE REVIEW

File: include/dlaf/sender/transform_mpi.h [hunk 1]
```diff
@@ -50,23 +66,25 @@ struct MPICallHelper {
...
     };
 
-    using result_type = decltype(pika::unwrapping(
-        std::move(f))(movePromiseGuard(dlaf::internal::getReferenceWrapper(ts))..., &req));
+    using result_type = decltype(pika::unwrapping(std::move(
+        f))(getPromiseGuardRef(movePromiseGuard(dlaf::internal::getReferenceWrapper(ts)))..., &req));
     if constexpr (std::is_void_v<result_type>) {
-      pika::unwrapping(std::move(f))(movePromiseGuard(dlaf::internal::getReferenceWrapper(ts))..., &req);
+      pika::unwrapping(std::move(
+          f))(getPromiseGuardRef(movePromiseGuard(dlaf::internal::getReferenceWrapper(ts)))..., &req);
```
Comment (msimberg) [inline include/dlaf/sender/transform_mpi.h:82]: Something like this:
```suggestion
      pika::unwrapping(std::move(
          f))(getPromiseGuardRef(dlaf::internal::getReferenceWrapper(ts))..., &req);
      consumePromiseGuard(ts...);
```
where `consumePromiseGuard` would be something like this:
```
template <typename T>
void consumePromiseGuard(T& t) {
  if constexpr (IsPromiseGuardCommunicator<std::decay_t<T>>::value) {
     auto t_local = std::move(t);
  }
```

**Review by msimberg — CHANGES_REQUESTED** [review]
@albestro just putting down in writing what we discussed on Friday. I think your idea of separating consuming the promise guard from the call to `f`. If you indeed had some idea for verifying this behaviour with a test I think that'd be great, even if it may be somewhat complicated...

File: include/dlaf/sender/transform_mpi.h (hunk 1 above)
Comment (msimberg) [reply include/dlaf/sender/transform_mpi.h]: That's a good point. I think adding `Communicator` to the name is a good idea.

**Review by msimberg — APPROVED** [review]

**Comment by msimberg** [comment]
bors try

**Comment by albestro** [comment]
bors try

File: test/unit/communication/test_transform_mpi.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  Communicator world(MPI_COMM_WORLD);
+  const IndexT_MPI rank = world.rank();
+
+  if (rank == 0) {
+    using dlaf::comm::internal::transformMPI;
+    using dlaf::internal::whenAllLift;
+
+    namespace ex = pika::execution::experimental;
+
+    common::Pipeline<Communicator> chain(world);
```
Comment (albestro) [inline test/unit/communication/test_transform_mpi.cpp:58]: Communicator `world` gets passed into Pipeline without being moved, and it gets used later for the signal message.

Copying a Communicator means copying the handle, so we end up using the same handle from different threads (in edge cases concurrently). It should be ok, since MPI calls should be thread-safe, moreover it is the `MPI_COMM_WORLD` communicator which is managed by MPI directly (so there are no additional caveats about lifetime management).

In case it is not, we may opt for cloning the Communicator on both ranks and using one for the `IRecv` communication and one for the signal part.

Let me know what you would prefer.

**Comment by albestro** [comment]
> If you indeed had some idea for verifying this behaviour with a test I think that'd be great, even if it may be somewhat complicated...

@msimberg @teonnik

@rasolca came up with an idea for the test, that we briefly implemented (and quickly checked some edge-cases by altering the implementation of `transformMPI`).

File: test/unit/communication/test_transform_mpi.cpp (hunk 2 above)
Comment (rasolca) [reply test/unit/communication/test_transform_mpi.cpp:58]: No need to duplicate in this case as only two messages that have a unique match are exchanged.

**Review by rasolca — APPROVED** [review]

**Comment by rasolca** [comment]
bors try

File: test/unit/communication/test_transform_mpi.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+    // "ensure" (by waiting a reasonable amount of time) that IRecv doesn't complete
+    std::this_thread::sleep_for(std::chrono::milliseconds(200));
+    EXPECT_FALSE(sent_guard);
+
+    // Note:
+    // At this point we checked if everything went as expected in the posting phase. Let's signal
+    // Rank 1 that it can send back the message that will unlock IRecv.
+    const int signal = 13;
+    MPI_Send(&signal, 1, MPI_INT, 1, 0, world);
```
Comment (teonnik) [inline test/unit/communication/test_transform_mpi.cpp:102]: I'm not sure I get this part. The `MPI_Send` is in standard mode, so it's implementation dependent whether the call will complete before the matching `MPI_Recv` is issued. The call may complete as soon as the data is buffered (but not yet sent). Perhaps use `MPI_Rsend`? this will start the `send` only when the matching `recv` is posted.

Comment (rasolca) [reply test/unit/communication/test_transform_mpi.cpp:102]: Not sure I understand the question. This MPI_Send is just triggering the other rank to send the message that match the receive posted with trasformMPI. When it returns doesn't influence the test.

The timeline in the following:
```
RANK 0:                                            |  Rank 1:
- Post MPI_Irecv via tranformMPI (l. 67)           | - MPI_Recv (l. 123)
- check that the comm pipeline has unlocked (92)   |    |
  (as the MPI_Irecv can be immediately posted)     |    |
- check that the continuation of the transformMPI  |    |
  hasn't been unlocked (l. 96)                     |    |
- MPI_Send (l. 102)        ------------------------|--> Ok
(Match with previous Irecv) <----------------------| - MPI_Send (l. 129)
- check that the continuation is now unlocked (111)
```
Note that all checks are performed with a delay, as certain operations might not execute immediately.

**Comment by rasolca** [comment]
bors try
